internal/cmd/gocdk/init.go

```diff
@@ -15,11 +15,101 @@
 package main
 
 import (
+	"bytes"
 	"context"
+	"flag"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
 
+	"gocloud.dev/internal/cmd/gocdk/internal/templates"
 	"golang.org/x/xerrors"
 )
 
 func init_(ctx context.Context, pctx *processContext, args []string) error {
-	return xerrors.New("not implemented")
+	f := newFlagSet(pctx, "init")
+	var modpath string
+	f.StringVar(&modpath, "module-path", "", "the module path for your project's go.mod file")
+	f.StringVar(&modpath, "m", "", "alias for --module-path")
+
+	if err := f.Parse(args); xerrors.Is(err, flag.ErrHelp) {
+		return nil
+	} else if err != nil {
+		return usagef("gocdk init: %w", err)
+	}
+
+	if f.NArg() != 1 {
+		return usagef("gocdk init [options] PATH")
+	}
+
+	path := pctx.resolve(f.Arg(0))
+	if modpath == "" {
+		var err error
+		modpath, err = inferModulePath(ctx, pctx, path)
+		if err != nil {
+			// TODO(clausti): return information about how to mitigate this error
+			// e.g. tell them to use --module-path to specify it
+			return xerrors.Errorf("gocdk init: %w", err)
+		}
+	}
+
+	// TODO(light): allow an existing empty directory, for some definition of empty
+	if _, err := os.Stat(path); err == nil {
+		return xerrors.Errorf("gocdk init: %s already exists", path)
+	} else if !os.IsNotExist(err) {
+		return xerrors.Errorf("gocdk init: %w", err)
+	}
+
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return xerrors.Errorf("gocdk init: %w", err)
+	}
+
+	tmplValues := struct {
+		ProjectName string
+		ModulePath  string
+	}{
+		ProjectName: filepath.Base(path),
+		ModulePath:  modpath,
+	}
+	for name, templateSource := range templates.InitTemplates {
+		tmpl := template.Must(template.New("").Parse(templateSource))
+		buf := new(bytes.Buffer)
+		if err := tmpl.Execute(buf, tmplValues); err != nil {
+			return xerrors.Errorf("gocdk init: %w", err)
+		}
+
+		fullPath := filepath.Join(path, filepath.FromSlash(name))
+		if err := os.MkdirAll(filepath.Dir(fullPath), 0777); err != nil {
+			return xerrors.Errorf("gocdk init: %w", err)
+		}
+		if err := ioutil.WriteFile(fullPath, buf.Bytes(), 0666); err != nil {
+			return xerrors.Errorf("gocdk init: %w", err)
+		}
+	}
+	return nil
+}
+
+// inferModulePath will check the default GOPATH to attempt to infer the module
+// import path for the project.
+func inferModulePath(ctx context.Context, pctx *processContext, path string) (string, error) {
```

Comment:
I'm having trouble understanding how this works. Maybe it's not important. Tests might provide some examples to make it more clear (maybe factor out the "infer" part from the "exec go" part).

For now, maybe just add a TODO to add tests for this.

Reply:
https://github.com/google/go-cloud/issues/2016